### PR TITLE
Add remove token cache item for a specific user

### PR DESCRIPTION
**Pull request by unpluggedk**
Add to public API that enables apps to delete all token cache items for a specific user.

Add few missing comments for other remove methods in the  ADKeychainTokenCache.h

### CODE REVIEW

File: ADAL/src/cache/ios/ADKeychainTokenCache.m [hunk 1]
```diff
@@ -451,6 +451,34 @@ - (BOOL)removeAllForUserId:(NSString * __nonnull)userId
...
+
+    for (ADTokenCacheItem *item in items)
+    {
+        if ([userId isEqualToString:[[item userInformation] userId]])
+        {
+            [self removeItem:item error:&adError];
+            if (adError)
+            {
+                deleteSuccessful = NO;
+                if (error)
```
Comment (partnerinflight) [inline ADAL/src/cache/ios/ADKeychainTokenCache.m]: CRITICAL Refactor this code to not nest more than 3 if|for|do|while|switch statements. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS134)

**Comment by partnerinflight** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: ADAL/src/cache/ios/ADKeychainTokenCache.m [hunk 2]
```diff
@@ -451,6 +451,34 @@ - (BOOL)removeAllForUserId:(NSString * __nonnull)userId
...
+    NSArray *items = [self allItems:nil];
+
+    ADAuthenticationError *adError = nil;
+
+    for (ADTokenCacheItem *item in items)
+    {
+        if ([userId isEqualToString:[[item userInformation] userId]])
+        {
+            [self removeItem:item error:&adError];
+            if (adError)
```
Comment (RPangrle) [inline ADAL/src/cache/ios/ADKeychainTokenCache.m]: should check bool return value, not adError

Comment (unpluggedk) [reply ADAL/src/cache/ios/ADKeychainTokenCache.m]: Refactoring, no need for adError and the deleteSuccessful boolean.
Previously, it was going through all items even though there was an error along the way.

File: ADAL/src/cache/ios/ADKeychainTokenCache.m [hunk 3]
```diff
@@ -427,28 +418,40 @@ - (BOOL)removeAllForUserId:(NSString * __nonnull)userId
...
+    AD_LOG_WARN_DICT(([NSString stringWithFormat:@"Removing all items for user <%@>", userId]), nil,
+                     (@{ @"operation" : @"removeAllForClientId:", @"userId" : userId }), nil);
+
+    NSArray *items = [self allItems:nil];
+
+    for (ADTokenCacheItem *item in items)
+    {
+        if ([userId isEqualToString:[[item userInformation] userId]])
+        {
+            if (![self removeItem:item error:error])
```
Comment (partnerinflight) [inline ADAL/src/cache/ios/ADKeychainTokenCache.m]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1066)

File: ADAL/src/cache/ios/ADKeychainTokenCache.m [hunk 4]
```diff
@@ -395,28 +395,19 @@ - (BOOL)removeAllForClientId:(NSString * __nonnull)clientId
...
     
-    ADAuthenticationError* adError = nil;
-    
     for (ADTokenCacheItem * item in items)
     {
         if ([clientId isEqualToString:[item clientId] ])
         {
-            [self removeItem:item error:&adError];
-            if (adError)
+            if (![self removeItem:item error:error])
```
Comment (partnerinflight) [inline ADAL/src/cache/ios/ADKeychainTokenCache.m]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1066)

File: ADAL/src/cache/ios/ADKeychainTokenCache.m [hunk 5]
```diff
@@ -427,28 +418,40 @@ - (BOOL)removeAllForUserId:(NSString * __nonnull)userId
...
-    ADAuthenticationError* adError = nil;
-    
     for (ADTokenCacheItem * item in items)
     {
         if ([userId isEqualToString:[[item userInformation] userId]]
             && [clientId isEqualToString:[item clientId]])
         {
-            [self removeItem:item error:&adError];
-            if (adError)
+            if (![self removeItem:item error:error])
```
Comment (partnerinflight) [inline ADAL/src/cache/ios/ADKeychainTokenCache.m]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=objc%3AS1066)

**Comment by partnerinflight** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: ADAL/src/cache/ios/ADKeychainTokenCache.m [hunk 6]
```diff
@@ -427,28 +416,36 @@ - (BOOL)removeAllForUserId:(NSString * __nonnull)userId
...
+            return NO;
+        }
+    }
+    return YES;
+}
+
+- (BOOL)removeAllForUserId:(NSString *)userId error:(ADAuthenticationError *__autoreleasing  _Nullable *)error
+{
+    AD_LOG_WARN_DICT(([NSString stringWithFormat:@"Removing all items for user <%@>", userId]), nil,
+                     (@{ @"operation" : @"removeAllForClientId:", @"userId" : userId }), nil);
```
Comment (oldalton) [inline ADAL/src/cache/ios/ADKeychainTokenCache.m]: nit: operation says removeAllForClientId :)

**Pull request merged**